File: NetFramework/Ribbon/VB/ControlHost/ControlHostLib/AxWebBrowserHost.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports C1.Win.C1Ribbon

''' <summary>
''' The Ribbon element that contains an ActiveX element (WebBrowser).
''' </summary>
Public Class AxWebBrowserHost
    Inherits RibbonControlHost

    Private _nextUrl As String
    Public Const XmlNodeName As String = "axWebBrowserHost"

    ''' <summary>
    ''' Creates the ActiveX control and its host element.
    ''' </summary>
    Public Sub New()
        MyBase.New(New AxSHDocVw.AxWebBrowser())
    End Sub

    ''' <summary>
    ''' <inheritdoc />
    ''' </summary>
    Public Overrides ReadOnly Property ControlHostNodeName As String
        Get
            Return XmlNodeName
        End Get
    End Property

    ''' <summary>
    ''' Gets the hosted AxWebBrowser control.
    ''' </summary>
    <Browsable(False)> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public ReadOnly Property AxWebBrowser() As AxSHDocVw.AxWebBrowser
        Get
            Return CType(Control, AxSHDocVw.AxWebBrowser)
        End Get
    End Property

    ''' <summary>
    ''' Navigates the browser to the specified URL.
    ''' </summary>
    Public Sub Navigate(ByVal url As String)
        If Not AxWebBrowser.IsHandleCreated Then
            _nextUrl = url
            AddHandler AxWebBrowser.HandleCreated, AddressOf AxWebBrowser_HandleCreated
        Else
            AxWebBrowser.Navigate(url)
        End If
    End Sub

    Private Sub AxWebBrowser_HandleCreated(ByVal sender As Object, ByVal e As EventArgs)
        RemoveHandler AxWebBrowser.HandleCreated, AddressOf AxWebBrowser_HandleCreated
        If _nextUrl IsNot Nothing Then
            AxWebBrowser.Navigate(_nextUrl)
            _nextUrl = Nothing
        End If
    End Sub

End Class

File: NetFramework/Ribbon/CS/ControlHost/ControlHostLib/AxWebBrowserHost.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using C1.Win.C1Ribbon;

namespace ControlHostLib
{
    /// <summary>
    /// The Ribbon element that contains an ActiveX element (WebBrowser).
    /// </summary>
    public class AxWebBrowserHost : RibbonControlHost
    {
        public const string XmlNodeName = "axWebBrowserHost";

        string _nextUrl;

        /// <summary>
        /// Creates the ActiveX control and its host element.
        /// </summary>
        public AxWebBrowserHost()
            : base(new AxSHDocVw.AxWebBrowser())
        {
        }

        /// <inheritdoc />
        public override string ControlHostNodeName
        {
            get { return XmlNodeName; }
        }

        /// <summary>
        /// Gets the hosted AxWebBrowser control.
        /// </summary>
        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public AxSHDocVw.AxWebBrowser AxWebBrowser
        {
            get { return (AxSHDocVw.AxWebBrowser)Control; }
        }

        /// <summary>
        /// Navigates the browser to the specified URL.
        /// </summary>
        public void Navigate(string url)
        {
            if (!AxWebBrowser.IsHandleCreated)
            {
                _nextUrl = url;
                AxWebBrowser.HandleCreated += new EventHandler(AxWebBrowser_HandleCreated);
            }
            else
                AxWebBrowser.Navigate(url);
        }

        private void AxWebBrowser_HandleCreated(object sender, EventArgs e)
        {
            AxWebBrowser.HandleCreated -= new EventHandler(AxWebBrowser_HandleCreated);
            if (_nextUrl != null)
            {
                AxWebBrowser.Navigate(_nextUrl);
                _nextUrl = null;
            }
        }
    }
}
